compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -516,101 +514,7 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
 
     val childTp = if (child.isTerm) child.termRef else child.typeRef
 
-    val resTp = instantiate(childTp, parent)(ctx.fresh.setNewTyperState())
-
-    if (!resTp.exists || !inhabited(resTp)) {
```

Comment:
Yes, it takes the bounds & variance of abstract type & type parameters  into consideration.

So it reports non-exhaustive warnings for the following code (it is possible that `T = Int | Boolean`):

```Scala
object O {
   sealed trait Expr[+T]
   case class IExpr(x: Int) extends Expr[Int]
   case class BExpr(b: Boolean) extends Expr[Boolean]

   def foo[T](x: Expr[T], y: Expr[T]) = (x, y) match {
      case (IExpr(_), IExpr(_)) => true
      case (BExpr(_), BExpr(_)) => false
   }
}
```

It does not report warning for the following code, due to nonvariance of the type parameter:

```Scala
object O {
  sealed trait Expr[T]
  case class BExpr(bool: Boolean) extends Expr[Boolean]
  case class IExpr(int: Int) extends Expr[Int]

  def join[T](e1: Expr[T], e2: Expr[T]): Expr[T] = (e1, e2) match {
    case (IExpr(i1), IExpr(i2)) => IExpr(i1 + i2)
    case (BExpr(b1), BExpr(b2)) => BExpr(b1 & b2)
  }
}
```

Reply:
Yes I did that in a separate commit because it requires some extra logic: this check from space takes a type as input and looks for unrealizable intersections inside, while the other check started from two types and checks if their intersection is unrealizable. The commits move a test back and forth in `/pending` to document what is handled by that code.